sql/core/src/main/java/org/apache/spark/sql/sources/v2/writer/DataSourceV2Writer.java

```diff
@@ -0,0 +1,71 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.writer;
+
+import org.apache.spark.annotation.InterfaceStability;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.sources.v2.DataSourceV2Options;
+import org.apache.spark.sql.sources.v2.WriteSupport;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * A data source writer that is returned by
+ * {@link WriteSupport#createWriter(StructType, SaveMode, DataSourceV2Options)}.
+ * It can mix in various writing optimization interfaces to speed up the data saving. The actual
+ * writing logic is delegated to {@link WriteTask} that is returned by {@link #createWriteTask()}.
+ *
+ * The writing procedure is:
+ *   1. Create a write task by {@link #createWriteTask()}, serialize and send it to all the
+ *      partitions of the input data(RDD).
+ *   2. For each partition, create a data writer with the write task, and write the data of the
+ *      partition with this writer. If all the data are written successfully, call
+ *      {@link DataWriter#commit()}. If exception happens during the writing, call
+ *      {@link DataWriter#abort()}. This step may repeat several times as Spark will retry failed
+ *      tasks.
+ *   3. Wait until all the writers/partitions are finished, i.e., either committed or aborted. If
+ *      all partitions are written successfully, call {@link #commit(WriterCommitMessage[])}. If
+ *      some partitions failed and aborted, call {@link #abort()}.
+ *
+ * Note that, data sources are responsible for providing transaction ability by implementing the
+ * `commit` and `abort` methods of {@link DataSourceV2Writer} and {@link DataWriter} correctly.
+ * The transaction here is Spark-level transaction, which may not be the underlying storage
+ * transaction. For example, Spark successfully write data to a Cassandra data source, but
+ * Cassandra may need some more time to reach consistency at storage level.
+ */
+@InterfaceStability.Evolving
+public interface DataSourceV2Writer {
+
+  /**
+   * Creates a write task which will be serialized and sent to executors. For each partition of the
+   * input data(RDD), there will be one write task to write the records.
+   */
+  WriteTask<Row> createWriteTask();
+
+  /**
+   * Commits this writing job with a list of commit messages. The commit messages are collected from
+   * all data writers for this writing job and are produced by {@link DataWriter#commit()}. This
+   * also means all the data are written successfully and all data writers are committed.
+   */
+  void commit(WriterCommitMessage[] messages);
+
+  /**
+   * Aborts this writing job because some data writers are failed to write the records and aborted.
+   */
+  void abort();
```

Comment:
Should this accept the commit messages for committed tasks, or will tasks be aborted?

I'm thinking of the case where you're writing to S3. Say a data source writes all attempt files to the final locations, then removes any attempts that are aborted. If the job aborts with some tasks that have already committed, then either this should have the option of cleaning up those files (passed in the commit message) or all of the tasks should be individually aborted. I'd prefer to have this abort clean up successful/committed tasks because the logic may be different.

Reply:
Let's say we have 400 partitions to write, and first 100 write tasks are successful and committed. If the 101st write task failed, there is no way to re-launch the previous 100 write tasks and abort them, we can only ask `DataSourceV2Writer` to do a job-level abort.

So you are right, this method should also take a list of commit messages for already-successful write tasks.

Reply:
After thinking it more, we can't do this. A spark job may crash without knowing which tasks success and which fail. So the `abort` here should be able to handle already committed tasks.

Reply:
Can you explain the situation you're talking about a bit more?

I think Spark should pass everything it can to the abort. I agree that the abort here should be flexible and a best-effort, but there are situations where it can't know how to roll back everything that tasks did without those commit messages. There may be cases where Spark can't guarantee whether a commit was complete or not, but where it can, it should pass those commit messages.

Say you had 100 successful write tasks of 400, but the executor died before it returned the 100th message (but after committing data) and that executor failure hit the max and Spark cancelled the job. Even though Spark has only 99 commit messages, it should still pass the ones that it can. In the S3 case, these are the only way the driver knows to delete the other 99 files. So the choice is between deleting 99 of 100 committed files, and leaving all 100 taking up space. I think rolling back 99 is much better than 0.

Reply:
Knowing what's being committed can provide a bit more information as to what is going on, and could be appreciated for that. The biggest issue I fear is loss of the driver itself, so *no* abort() call is made. A strategy for cleaning up from that would be good, even though its primarily one of bringing up a new writer and saying "cleanup this mess a previous instance left". Looking the S3 example, my strategy there is/would be: identify all pending commits, abort them, then clean up the dest dir.